### PR TITLE
Distributed testing support [not ready for merge]

### PR DESCRIPTION
**Pull request by dabrahams**
## Run your Swift tests on a cluster.

In my fairly crude setup with 4 machines, testing time was reduced by 1/3.  I used thunderbolt bridge networking for the cluster to reduce the impact of file sharing.  Further testing is ongoing.
### Quickstart

Get two or more machines.  One is the master and the others are slaves.
- On the master, create a directory and share it with all the slaves. You want it to have the same path on all the machines, so put it under `/Volumes/` or appropriately control where you mount it on the slaves.
- On the master, create a `~/.distcc/hosts` file listing all the hosts by name or IP address and specifying build parallelism, e.g.
  
  ```
   --randomize
   localhost/10,lzo,cpp # usually 5/4 * number of cores works well
   slave-1/10,lzo,cpp
   slave-2/5,lzo,cpp
  ```
- Be able to ssh into the slaves from the master without giving a password (e.g. using `.ssh/authorized_keys`)
- Check out this branch of swift into the shared directory (or merge it into your branch)
- `$ cd swift/`
- `$ utils/update-checkout --clone[-with-ssh] # get swift's dependencies`
- add `--lit-args='--distribute -sv'` to your usual invocation of `utils/build-script`
- build and test normally.

### CODE REVIEW

**Comment by dabrahams** [comment]
@swift-ci Please smoke test

File: utils/lit/lit/ProgressBar.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+    output; and adjusts to the width of the terminal.
+    """
+    BAR = '%s${GREEN}[${BOLD}%s%s${NORMAL}${GREEN}]${NORMAL}%s'
+    HEADER = '${BOLD}${CYAN}%s${NORMAL}\n\n'
+        
+    def __init__(self, term, header, useETA=True):
+        self.term = term
+        if not (self.term.CLEAR_EOL and self.term.UP and self.term.BOL):
+            raise ValueError("Terminal isn't capable enough -- you "
+                             "should use a simpler progress dispaly.")
```
Comment (karwa) [inline utils/lit/lit/ProgressBar.py]: typo

**Comment by gottesmm** [comment]
Cool!

**Comment by karwa** [comment]
This is great! Hopefully we can adapt this when cross-compiling.

File: utils/lit/lit/ProgressBar.py (hunk 1 above)
Comment (dabrahams) [inline utils/lit/lit/ProgressBar.py]: Thanks, but this was verbatim-imported from LLVM and I'm going to limit my changes to what's needed for distributed build support; you should file your comment/bug against the llvm source.

**Comment by dabrahams** [comment]
closing in favor of #4611

**Pull request closed without merging**